base_layer/p2p/src/services/liveness/service.rs

```diff
@@ -0,0 +1,223 @@
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use super::{error::LivenessError, state::LivenessState, LivenessRequest, LivenessResponse};
+use crate::{
+    services::{comms_outbound::CommsOutboundHandle, ServiceHandles, ServiceName},
+    tari_message::{NetMessage, TariMessageType},
+};
+use futures::{
+    future::{self, Either},
+    Future,
+    Poll,
+};
+use serde::{Deserialize, Serialize};
+use std::{convert::TryInto, sync::Arc};
+use tari_comms::{
+    message::{Message, MessageError, MessageFlags},
+    outbound_message_service::BroadcastStrategy,
+    types::CommsPublicKey,
+};
+use tower_service::Service;
+
+/// The PingPong message
+#[derive(Serialize, Deserialize)]
+pub enum PingPong {
+    Ping,
+    Pong,
+}
+
+impl TryInto<Message> for PingPong {
+    type Error = MessageError;
+
+    fn try_into(self) -> Result<Message, Self::Error> {
+        Ok((TariMessageType::new(NetMessage::PingPong), self).try_into()?)
+    }
+}
+
+/// Service responsible for testing Liveness for Peers.
+///
+/// Very basic global ping and pong counter stats are implemented. In future,
+/// peer latency and availability stats will be added.
+pub struct LivenessService {
+    state: Arc<LivenessState>,
+    handles: Arc<ServiceHandles>,
+    oms_handle: Option<CommsOutboundHandle>,
+}
+
+impl LivenessService {
+    pub fn new(state: Arc<LivenessState>, handles: Arc<ServiceHandles>) -> Self {
+        Self {
+            state,
+            handles,
+            oms_handle: None,
+        }
+    }
+
+    /// Get the CommsOutboundHandle.
+    ///
+    /// This function will panic if the handle does not exist.
+    fn oms_handle(&mut self) -> &mut CommsOutboundHandle {
+        match self.oms_handle {
+            Some(ref mut h) => h,
+            None => {
+                // Memoize the handle for later use
+                self.oms_handle = Some(
+                    self.handles
+                        .get_handle::<CommsOutboundHandle>(&ServiceName::CommsOutbound)
+                        .expect("LivenessService requires CommsOutboundHandle"),
+                );
+                self.oms_handle()
+            },
+        }
+    }
+
+    fn send_ping(
+        &mut self,
+        pub_key: CommsPublicKey,
+    ) -> impl Future<Item = Result<LivenessResponse, LivenessError>, Error = ()>
```

Comment:
Shouldn't / Can this be flattened to 
`Future<Item = LivenessResponse, Error = LivenessError>`

Reply:
Now that I'm a bit further down the line, I think it's worth looking at - there are other errors that can occur that are not related to the service or the request being made. Dealing with those errors separately is easier. Still, not sure it's the right way to go about it since the unrelated errors will probably be due to some 'coding error' (incorrect runtime setup/ unclean shutdown etc) rather than a legitimate error.